embedded-service/src/fw_update.rs

```diff
@@ -0,0 +1,72 @@
+//! Module for general firmware updates
+use core::future::Future;
+
+/// FW update error kind
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum ErrorKind {
+    /// The device is not in FW update mode
+    InvalidMode,
+    /// FW update Content was invalid
+    InvaildContent,
+    /// Full FW update contents were not received
+    Incomplete,
+    /// The operation did not complete in the expected time
+    Timeout,
+    /// Error communicating with the device due to a bus error
+    Bus,
+    /// Generic failure
+    Failed,
+}
+
+/// FW update error trait
+pub trait Error {
+    /// Map the error to a generic error
+    fn kind(&self) -> ErrorKind;
+}
+
+impl Error for ErrorKind {
+    fn kind(&self) -> ErrorKind {
+        *self
+    }
+}
+
+/// Convienent error type for devices that are on a bus
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum BusError<B> {
+    /// Bus error
+    Bus(B),
+    /// Generic error
+    General(ErrorKind),
+}
+
+impl<B> Error for BusError<B> {
+    fn kind(&self) -> ErrorKind {
+        match self {
+            BusError::Bus(_) => ErrorKind::Bus,
+            BusError::General(e) => *e,
+        }
+    }
+}
+
+/// General firmware update trait
+pub trait FwUpdate {
+    /// Error type
+    type Error: Error;
+
+    /// Get the current firmware version
+    fn get_active_fw_version(&self) -> impl Future<Output = Result<u32, Self::Error>>;
+
+    /// Begin a new firmware update
+    fn start_fw_update(&mut self) -> impl Future<Output = Result<(), Self::Error>>;
+
+    /// Terminate the firmware update after a failure
+    fn abort_fw_update(&mut self) -> impl Future<Output = Result<(), Self::Error>>;
```

Comment:
Can you explain why this is needed? From what I can tell, this will be a situation where this function *must* be called any time `start_fw_update()` fails. If that's the case, why have this method at all? Could just require the implementer to handle the errors and place the flash back in a working state as if the update had not been started.